core/src/main/java/hudson/model/Job.java

```diff
@@ -1101,50 +1089,68 @@ public FeedItem(ChangeLogSet.Entry e, int idx) {
                 getUrl() + "changes",
-                entries, new FeedAdapter<FeedItem>() {
-                    public String getEntryTitle(FeedItem item) {
-                        return "#" + item.getBuild().number + ' ' + item.e.getMsg() + " (" + item.e.getAuthor() + ")";
+                entries, new FeedAdapter<RssFeedItem>() {
+                    public String getEntryTitle(RssFeedItem item) {
+                        return "#" + item.getRun().number + ' ' + item.getEntry().getMsg() + " (" + item.getEntry().getAuthor() + ")";
                     }
 
-                    public String getEntryUrl(FeedItem item) {
-                        return item.getBuild().getUrl() + "changes#detail" + item.idx;
+                    public String getEntryUrl(RssFeedItem item) {
+                        return item.getRun().getUrl() + "changes#detail" + item.getIndex();
                     }
 
-                    public String getEntryID(FeedItem item) {
+                    public String getEntryID(RssFeedItem item) {
                             return getEntryUrl(item);
                         }
 
-                    public String getEntryDescription(FeedItem item) {
+                    public String getEntryDescription(RssFeedItem item) {
                         StringBuilder buf = new StringBuilder();
-                        for (String path : item.e.getAffectedPaths())
+                        for (String path : item.getEntry().getAffectedPaths())
                             buf.append(path).append('\n');
                         return buf.toString();
                     }
 
-                    public Calendar getEntryTimestamp(FeedItem item) {
-                            return item.getBuild().getTimestamp();
+                    public Calendar getEntryTimestamp(RssFeedItem item) {
+                            return item.getRun().getTimestamp();
                         }
 
-                    public String getEntryAuthor(FeedItem entry) {
+                    public String getEntryAuthor(RssFeedItem entry) {
                         return JenkinsLocationConfiguration.get().getAdminAddress();
                     }
                 },
                 req, rsp);
     }
 
+    /**
+     * Gets Rss Feed entries.
+     *
+     * @param limit limits the number of entries, in case of 0 unlimited.
+     * 
+     * @return The entries (never null).
+     */
+    @NonNull
+    private List<RssFeedItem> getEntries(int limit) {
+        List<RssFeedItem> result = new ArrayList<>();
+        for (RunT r = getLastBuild(); r != null; r = r.getPreviousBuild()) {
+            int idx = 0;
+            if (r instanceof RunWithSCM) {
+                for (ChangeLogSet<? extends ChangeLogSet.Entry> c : ((RunWithSCM<?,?>) r).getChangeSets()) {
+                    for (ChangeLogSet.Entry e : c) {
+                        result.add(new RssFeedItem(e, idx++));
+                        if (limit != 0 && result.size() >= limit) {
```

Comment:
I think this is the one line that actually needed to be changed. It is hard to review a diff that includes refactoring of existing code.